2017 Robot/src/org/usfirst/frc/team4999/lights/Packet.java

```diff
@@ -0,0 +1,153 @@
+package org.usfirst.frc.team4999.lights;
+
+import java.nio.ByteBuffer;
+
+/**
+ * A command sent to the arduino to light up a pixel or area of pixels with a certain color.
+ * Follows <a href="https://github.com/momentumfrc/2017Steamworks/wiki/Light-strings-on-the-Robot">this</a> structure.
+ * @author jordan
+ *
+ */
+public class Packet {
+	// The max size of a packet is 16 bytes
+	// Before a packet is sent, B_BUFFER is filled with that packet's data. B_BUFFER is then passed to the I2C writeBulk() method
+	private static final ByteBuffer B_BUFFER = ByteBuffer.allocateDirect(16);
+	
+	// All according to specification
+	private final byte DISPLAY_FRAME = 0x01;
+	private final byte SET_SINGLE = 0x02;
+	private final byte SET_RUN = 0x03;
+	private final byte SET_STRIDE = 0x04;
+	
+	
+	private byte command;
+	private byte[] data;
+	
+	/**
+	 * Creats a DISPLAY_FRAME packet
+	 */
+	public Packet() {
+		command = DISPLAY_FRAME;
+		data = new byte[0];
+	}
+	
+	/**
+	 * Sets color of a single pixel
+	 * @param address pixel to set
+	 * @param color color to set to
+	 * @see #Packet(int, Color, int) 
+	 * @see #Packet(int, Color, int, int)
+	 */
+	public Packet(int address, Color color) {
+		command = SET_SINGLE;
+		Color dimColor = BrightnessFilter.dimColor(color);
```

Comment:
Better, but I still think that the BrightnessFilter doesn't belong in the Packet class.  Packet has to do with packaging the command and data for transport, it shouldn't know anything about the data it's packaging, nor how other layers will pass the package along.  This is part of why I was suggesting the use of the Factory pattern to generate packets which separates the Packet from the packet generation.  I'm happy to demonstrate.

Reply:
I'm not quite sure what you mean by 'the Factory pattern'? I do agree that the Packet class probably should just be holding data, and not modifying it. Do you mean another class which has different methods that generate Packet objects?

Reply:
Here's the definition: https://en.wikipedia.org/wiki/Factory_method_pattern

Although it goes a bit abstract in the middle, so you're probably looking for a more concrete example.

In our specific case, there are two common ways to get this done.  You can have one class that knows how to properly generate other classes, including selection of what specific subclasses should be used for what situations.  If you go that route, I'd be looking for a series of APIs like this:
* Create a command packet:  
`Packet command = Commands.paintPixel(0, color);`  
`Packet command = Commands.paintPixels(0, color, 1);`  
`Packet command = Commands.paintPixelsWithStride(0, color, 1, 1);`

An alternative would be to simply use Packet as the base class for all Packet types.  Then the different commands would all be subclassed off of that like this:
* `Packet command = new PaintPixel(0, color);`
* `Packet command = new PaintPixels(0, color, 1);`
* `Packet command = new PaintPixelsWithStride(0, color, 1, 1);`

It may seem simpler at this time to choose the direct construction method, and that's likely to be the case.  The thing to watch out for is if there is any specific knowledge needed to create the painting commands, then a factory becomes much more useful since it's the central repository for knowledge of how to make the things.